src/librustc/hir/lowering.rs

```diff
@@ -539,73 +598,139 @@ impl<'a> LoweringContext<'a> {
-                    })
-                }).collect(),
-            None => Vec::new(),
-        };
+                // Add a definition for the in-band lifetime def
+                self.resolver.definitions().create_def_with_parent(
+                    parent_id.index,
+                    def_node_id,
+                    DefPathData::LifetimeDef(str_name),
+                    DefIndexAddressSpace::High,
+                    Mark::root(),
+                    span,
+                );
 
-        params.extend(in_band_ty_params.into_iter().map(|tp| hir::GenericParam::Type(tp)));
+                hir::GenericParam::Lifetime(hir::LifetimeDef {
+                    lifetime: hir::Lifetime {
+                        id: def_node_id,
+                        span,
+                        name: hir_name,
+                    },
+                    bounds: Vec::new().into(),
+                    pure_wrt_drop: false,
+                    in_band: true,
+                })
+            })
+            .chain(
+                in_band_ty_params
+                    .into_iter()
+                    .map(|tp| hir::GenericParam::Type(tp)),
+            )
+            .collect();
 
         (params, res)
     }
 
+    /// When there is a reference to some lifetime `'a`, and in-band
+    /// lifetimes are enabled, then we want to push that lifetime into
+    /// the vector of names to define later. In that case, it will get
+    /// added to the appropriate generics.
+    fn maybe_collect_in_band_lifetime(&mut self, span: Span, name: Name) {
+        if !self.is_collecting_in_band_lifetimes {
+            return;
+        }
+
+        if self.in_scope_lifetimes.contains(&name) {
+            return;
+        }
+
+        let hir_name = hir::LifetimeName::Name(name);
+
+        if self.lifetimes_to_define
+            .iter()
+            .any(|(_, lt_name)| *lt_name == hir_name)
+        {
+            return;
+        }
+
+        self.lifetimes_to_define.push((span, hir_name));
+    }
+
+    /// When we have either an elided or `'_` lifetime in an impl
+    /// header, we convert it to
+    fn collect_fresh_in_band_lifetime(&mut self, span: Span) -> hir::LifetimeName {
+        assert!(self.is_collecting_in_band_lifetimes);
+        let index = self.lifetimes_to_define.len();
+        let hir_name = hir::LifetimeName::Fresh(index);
+        self.lifetimes_to_define.push((span, hir_name));
+        hir_name
+    }
+
     // Evaluates `f` with the lifetimes in `lt_defs` in-scope.
     // This is used to track which lifetimes have already been defined, and
     // which are new in-band lifetimes that need to have a definition created
     // for them.
-    fn with_in_scope_lifetime_defs<T, F>(
+    fn with_in_scope_lifetime_defs<'l, T, F>(
         &mut self,
-        lt_defs: &[LifetimeDef],
-        f: F
-    ) -> T where F: FnOnce(&mut LoweringContext) -> T
+        lt_defs: impl Iterator<Item = &'l LifetimeDef>,
```

Comment:
Huh-- this is interesting. Seems like this would probably trigger https://github.com/rust-lang/rust/issues/44752, but it's an error to omit the lifetime. We should probably be able to make argument-position `impl Trait` elided lifetimes "just work", but it's backcompat to fix, so I'm not super concerned about it.

Reply:
> We should probably be able to make argument-position impl Trait elided lifetimes "just work", but it's backcompat to fix, so I'm not super concerned about it.

Yeah, that was curious, but I came to the same conclusion. I think clearly they should work as a fresh lifetime (but I feel the same about `'_` in where clauses).

Reply:
Should probably file an issue about it.

Reply:
Filed https://github.com/rust-lang/rust/issues/49287